tests/structured_conf/test_structured_config.py

```diff
@@ -1051,3 +1047,100 @@ def test_update_userlist(
         else:
             with pytest.raises(ValidationError):
                 OmegaConf.update(cfg, "list", update_value, merge=True)
+
+    def test_union_set_valid_value(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.Book)
+        value = ["dude1", "dude2"]
+        cfg.author = value
+        assert cfg.author == value
+        assert isinstance(cfg._get_node("author"), UnionNode)
+
+    def test_union_set_valid_value_nested(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.Shelf)
+        value = module.Book(author="foo")
+        cfg.content = value
+        assert cfg.content == value
+        assert isinstance(cfg._get_node("content"), UnionNode)
+
+    def test_union_set_valid_value_nested_list(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.Shelf2)
+        value = [module.Book(), module.Book()]
+        cfg.content = value
+        assert cfg.content == value
+        assert isinstance(cfg._get_node("content"), UnionNode)
+
+    def test_union_set_invalid_value(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.Book)
+        value = module.Book()
+        with pytest.raises(ValidationError):
+            cfg.author = value
+
+    def test_optional_union_create(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.OptionalBook(author=None))
+        assert isinstance(cfg._get_node("author"), UnionNode)
+        assert (
+            cfg._get_node("author")._metadata.ref_type
+            == Union[str, List[str], type(None)]
+        )
+        assert cfg._get_node("author")._metadata.optional is True
```

Comment:
Minor simplification?
```suggestion
        assert cfg._get_node("author")._is_optional()
```